### PR TITLE
PIV updates

### PR DESCRIPTION
**Pull request by FFincannon**

### CODE REVIEW

**Comment by FFincannon** [comment]
There are 8 items in the conf.json file for the PIV Guides, but 10 .adoc files in the Guides folder - the EJBCA one and Generating Keys Using OpenSSL are missing.

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc [hunk 1]
```diff
@@ -1,76 +1,59 @@
...
 * OpenSSH
-** If you are using OSX El Capitan (10.11) or earlier, for ssh-agent to work a newer OpenSSH than is delivered with the system; macOS Sierra (10.12) contains a compatible version
+* For OS X, iOS 10.13 and later are supported
 
 [NOTE]
-The following example assume that you have not yet changed the management key. If you have changed the management key, add `--key` to the `yubico-piv-tool -a import-certificate` command below.
+RSA 4096-bit keys are not currently supported due to a limitation in the PIV spec: https://github.com/Yubico/yubico-piv-tool/issues/58
+
+[NOTE]
+The examples given in the following steps assume that you have not yet changed the management key. If you have changed the management key, add `--key` to the `yubico-piv-tool -a import-certificate` command below.
```
Comment (aveenismail) [inline content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Maybe add a sentence saying that we strongly recommend changing the management key, or that we strongly discourage users from keeping the default management key

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc [hunk 2]
```diff
@@ -1,76 +1,59 @@
...
+*Step 3*: Load the certificate:
 
    $ yubico-piv-tool -a import-certificate -s 9a -i cert.pem
 
-4. Find out where OpenSC has installed the pkcs11 module.
-
-  * For OS X with binary installation this is typically in `/Library/OpenSC/lib/`. Homebrew users can use:
-
-  $ sudo ln `brew list opensc | grep lib/opensc-pkcs11.so` /usr/local/lib/opensc-pkcs11.so
+*Step 4*: Find out where OpenSC has installed the pkcs #11 module. For OS X with binary installation, this is typically in `/Library/OpenSC/lib/`. For a Debian-based system, this is typically in `/usr/lib/x86_64-linux-gnu/`. From now on, we will call this location `$OPENSC_LIBS`.
```
Comment (aveenismail) [inline content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Now that we have a functional pkcs11 module (the ykcs11), we can use that instead of OpenSC's. For my debian based system, the ykcs11 module ends up in /usr/local/lib/libykcs11.so (/usr/lib/libykcs11.so is also a common location). On my colleagues macos, it is in /usr/local/lib/libykcs11.dylib

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc [hunk 3]
```diff
@@ -1,76 +1,59 @@
...
 
-  * For a Debian based system this is typically in `/usr/lib/x86_64-linux-gnu/`
-+
-After this we'll call this location `$OPENSC_LIBS`
-
-5. Export the public key in correct format for ssh and once you got it,
-add it to authorized_keys on the target system.
+*Step 5*: Export the public key in correct format for SSH and then add it to authorized_keys on the target system.
 
    $ ssh-keygen -D $OPENSC_LIBS/opensc-pkcs11.so -e
```
Comment (aveenismail) [inline content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Change "$OPENSC_LIBS/opensc-pkcs11.so" to XXX/libykcs11.so

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc [hunk 4]
```diff
@@ -1,76 +1,59 @@
...
 [NOTE]
-The command will export all keys stored on the YubiKey Neo.
-Hopefully it will keep the slot order so it should be not hard to guess which
-is the public key associated with your targeted private key.
+The command will export all keys stored on the YubiKey. The slot order should remain the same, thereby facilitating identification of the public key associated with your targeted private key.
 
-6. Authenticate to the target system using the new key:
+*Step 6*: Authenticate to the target system using the new key:
 
    $ ssh -I $OPENSC_LIBS/opensc-pkcs11.so user@remote.example.com
```
Comment (aveenismail) [inline content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Change "$OPENSC_LIBS/opensc-pkcs11.so" to XXX/libykcs11.so

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc [hunk 5]
```diff
@@ -1,76 +1,59 @@
...
 
-7. This can also be setup to work with ssh-agent: (Optional)
+*Step 7*: (Optional) This can also be set up to work with ssh-agent:
 
    $ ssh-add -s $OPENSC_LIBS/opensc-pkcs11.so
```
Comment (aveenismail) [inline content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Change "$OPENSC_LIBS/opensc-pkcs11.so" to XXX/libykcs11.so

File: content/WebAuthn/Supporting_U2F_or_FIDO2_Security_Keys_on_iOS_or_iPadOS/Security_Key_Compatibility.adoc [hunk 6]
```diff
@@ -19,7 +19,7 @@ Compatible U2F/FIDO2 Security Keys that currently work with iOS/iPadOS 13.3 devi
 |YubiKey 4                    |USB-A*                                                          |USB-A*                                     |USB-A*
 |========================
 
-* When used with an official Apple or high-end USB adapter
+\* When used with an official Apple or high-end USB adapter
```
Comment (aveenismail) [inline content/WebAuthn/Supporting_U2F_or_FIDO2_Security_Keys_on_iOS_or_iPadOS/Security_Key_Compatibility.adoc]: Is this change intentional?

File: content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc [hunk 7]
```diff
@@ -0,0 +1,38 @@
+== Securing SSH with OpenPGP or PIV
+SSH supports multiple types of asymmetric certificates when implementing secure communication with remote systems, including OpenPGP format and X.509 (PIV). The YubiKey supports both certificate formats, but typically deployment considerations or the particular use case will dictate the format.
```
Comment (aveenismail) [inline content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc]: Thinking of OpenPGP and PIV as certificate formats is a bit confusing. Not sure that OpenPGP even have certificates, actually. 

Can we write something along the lines of the following?

Secure communication with remote systems with the help of a YubiKey can be done either with the YubiKey’s PIV application using, for example, OpenSSH, or with the YubiKey’s PGP application. Typically, deployment considerations or the particular use case will dictate what application to use

Comment (Yubi-David) [reply content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc]: @FFincannon please check my grammar, but how does this sound (heavily based off of Aveen's feedback):

Secure communication with remote systems via SSH can be protected using key-based authentication with the user's key residing on a physical YubiKey device. The YubiKey offers two options when securing a user's SSH key data: OpenPGP and PIV. Typically, deployment considerations or the particular use case will dictate what application to use.

beyond username with the help of a YubiKey can be done either with the YubiKey’s PIV application using, for example, OpenSSH, or with the YubiKey’s PGP application. Typically, deployment considerations or the particular use case will dictate what application to use

Comment (qpernil) [reply content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc]: A key point to help someone choose:
If you use the OpenPGP option for SSH then SSH will use the same key as GIT would use to sign a commit (since GIT uses  the gpg command to sign). However, the OpenPGP option is more work to set up properly (ssh will speak to the gpg agent) so the PIV option would probably be easier for anyone who doesn't use GIT or who actually wants to use a different key for SSH and GIT. 

Furthermore (maybe repeating stuff you already know) for the PIV option one my use the opensc pkcs11 module, or our new ykcs11 module. Our ykcs11 is a lighter software package, so would probably be an easier path for someone who doesn't already have the opensc software installed. (The same goes for OpenPGP vs PIV as well, if you already have gpg working)

Comment (qpernil) [reply content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc]: So for me as a developer the OpenPGP path is the natural one, I need gpg working to be able to use git, and I only have to upload a single public key to github, which it uses to both authenticate my ssh access (used for git pushing) and for verifying the signatues on my git commits.

Comment (qpernil) [reply content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc]: Another detail: ykcs11 gives access to all 24 slots on the PIV application, while opensc only exposes 4. Additionally, the attestation certificates are only available with ykcs11.

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc (hunk 1 above)
Comment (FFincannon) [reply content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Done. And when the recommendation is this strong, emphasizing it as you did seems a good idea: 
[NOTE]
We strongly recommend changing the management key; keeping the default management key is explicitly discouraged. The examples given in the following steps assume that you have not yet changed the management key. If you have changed the management key, add `--key` to the `yubico-piv-tool -a import-certificate` command below.

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc (hunk 2 above)
Comment (FFincannon) [reply content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Done. And I removed the OpenSC as a prerequisite, too, adding why it is no longer needed.

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc (hunk 3 above)
Comment (FFincannon) [reply content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Line changed to:
 $ ssh-keygen -D XXX/libykcs11.so -e

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc (hunk 4 above)
Comment (FFincannon) [reply content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Line is now
   $ ssh -I XXX/libykcs11.so user@remote.example.com

File: content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc (hunk 5 above)
Comment (FFincannon) [reply content/PIV/Guides/SSH_with_PIV_and_PKCS11.adoc]: Line is now
 $ ssh-add -s XXX/libykcs11.so

File: content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc (hunk 7 above)
Comment (FFincannon) [reply content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc]: Even as a non-developer, I like to have a few concrete reasons for choosing one way or another. The new introductory paragraph reads:

Securing communication with remote systems via SSH can be done by using key-based authentication with the user's key residing on a physical YubiKey. With a YubiKey, the user's SSH key data can be secured using OpenPGP and/or PIV. Typically, deployment considerations or the particular use case will dictate which application to use.

If you use the OpenPGP option for SSH, then SSH will use the same key as Git would use to sign a commit (since Git uses the GnuPG command to sign). However, the OpenPGP option is more complicated to set up properly (SSH will speak to the GnuPG agent) so the PIV option would probably be easier for anyone who does not use Git or who does not want to use the same key for both SSH and Git. Furthermore, for the PIV option, you can use the opensc-pkcs11 module, or our new ykcs11 module. Our ykcs11 is a lighter software package, so it would probably offer an easier path for someone who does not already have the OpenSC software installed. (The same goes for OpenPGP vs PIV as well, if you already have GnuPG working).

For a developer the OpenPGP path is typically the natural one, because GnuPG needs to be working to enable the use of Git, and it is only necessary to upload a single public key to GitHub, which it then uses both to authenticate SSH access (used for Git pushing) and for verifying the signatures on Git commits.

File: content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+In larger organizations, OpenPGP’s ability to run without a greater framework also can hinder the uniform application of security policies. Without a central authority enforcing a standard set of attributes during key generation, there is a risk of a user unintentionally generating a weak key or setting an expiration date too far in the future or too soon. Again, this can be offset via training and documentation, but this mitigation strategy assumes users read the documentation.
+
+While OpenPGP is widely used for signing and encryption services, it is not commonly used for centrally managed authentication. When using certificate-based authentication, the lack of a framework to support a credential authority means OpenPGP is not supported. This and the fact that OpenPGP on a smart card is limited to a single cert and key (split into 3 sub-keys), credential management services and certificate authorities have not added native support for OpenPGP credentials.
+
+=== PIV / X.509
```
Comment (dainnilsson) [inline content/PIV/Guides/Securing_SSH_with_OpenPGP_or_PIV.adoc]: Everything that talks about X509 certificates in the context of SSH here is confusing to me. SSH uses its own (optional) certificate format which has nothing to do with X509. Regardless of if PIV or OpenPGP is used for SSH, only the cryptographic key pair is used, any certificates are disregarded. While potentially you could connect PKI to SSH authorization via the public key, there doesn't exist a widespread solution which does this that I'm aware of. Using SSH certificates (which are not X509) can be done with either PIV or OpenPGP - it doesn't matter.
